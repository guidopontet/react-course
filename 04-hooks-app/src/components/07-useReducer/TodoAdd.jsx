import React from 'react'

import useForm from '../../hooks/useForm'

export const TodoAdd = ({ handleAddTodo }) => {
  const [ { description }, handleInputChange, resetForm ] = useForm({
    description: '',
  })

  const handleSubmit = (e) => {
    e.preventDefault();

    if (description.trim().length < 1 ) { return; }

    const newTodo = {
      id: new Date().getTime(),
      description: description,
      done: false,
    };

    const action = {
      type: 'ADD',
      payload: newTodo,
    }

    handleAddTodo(action);
    resetForm();
  }

  return (
    <>
      <h4>Agregar TODO</h4>
      <hr />

      <form onSubmit={ handleSubmit }>
        <input type="text"
                className='form-control'
                name='description'
                placeholder='Ingresa la tarea'
                autoComplete='off'
                value={ description }
                onChange={ handleInputChange }/>

        <button type='submit'
                className='btn btn-primary btn-block mt-3'>
          Agregar
        </button>
      </form>
    </>
  )
}
