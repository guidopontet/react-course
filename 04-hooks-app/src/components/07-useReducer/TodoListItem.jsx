import React from 'react'

export const TodoListItem = ({ todo, index, handleToggle, handleDelete}) => {
  return (
    <li key={ todo.id }
        className='list-group-item'>
      <p className={ todo.done ? 'complete' : '' }
          onClick={ () => handleToggle(todo) }>
        { index + 1 }. { todo.text }
      </p>

      <button className='btn btn-danger'
              onClick={ () => handleDelete(todo) }>Borrar</button>
    </li>
  )
}
