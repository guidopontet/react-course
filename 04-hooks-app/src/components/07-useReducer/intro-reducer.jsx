const initialState = [{
  id: 1,
  todo: "Comprar pan",
  done: false,
}];

const todoReducer = (state = initialState, action) => {
  if (action?.type === "ADD_TODO") {
    return [...state, action.payload];
  }

  return state;
}

let todos = todoReducer();

const newTodo = {
  id: 2,
  todo: "Comprar pasta",
  done: false,
};

const addTodoAction = {
  type: "ADD_TODO",
  payload: newTodo,
}

todos = todoReducer(todos, addTodoAction);

console.log(todos);
