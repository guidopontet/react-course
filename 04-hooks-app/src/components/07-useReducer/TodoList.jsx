import React from 'react'

import { TodoListItem } from './TodoListItem';

export const TodoList = ({ todos, handleDelete, handleToggle }) => {
  return (
    <ul className='list-group list-group-flush'>
      {
        todos.map((todo, i) => {
          return (
            <TodoListItem todo={ todo }
                          index={ i }
                          key={ todo.id }
                          handleDelete= { handleDelete }
                          handleToggle={ handleToggle }/>
          );
        })
      }
    </ul>
  )
}
