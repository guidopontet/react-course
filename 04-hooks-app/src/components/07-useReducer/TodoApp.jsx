import React, { useEffect, useReducer } from 'react';

import { TodoAdd } from './TodoAdd';
import { TodoList } from './TodoList';
import { todoReducer } from './todoReducer';

import './styles.css';

const init = () => {
  return JSON.parse(localStorage.getItem('todos')) || [];
}

export const TodoApp = () => {
  const [ todos, dispatch ] = useReducer(todoReducer, [], init);

  useEffect(() => {
    localStorage.setItem('todos', JSON.stringify(todos));
  }, [todos]);

  const handleDelete = (todo) => {
    const action = {
      type: 'DELETE',
      payload: todo.id,
    }

    dispatch(action);
  }

  const handleToggle = (todo) => {
    const action = {
      type: 'TOGGLE',
      payload: todo.id,
    }

    dispatch(action);
  }

  const handleAddTodo = ((action) => {
    dispatch(action);
  })

  return (
    <div className='container my-3'>
      <h1>TodoApp ({ todos.length })</h1>
      <hr />

      <div className='row'>
        <div className='col-7'>
          <TodoList todos={ todos }
                    handleToggle={ handleToggle }
                    handleDelete= { handleDelete } />
        </div>
        <div className='col-5'>
          <TodoAdd handleAddTodo={ handleAddTodo }/>
        </div>
      </div>
    </div>
  )
}
