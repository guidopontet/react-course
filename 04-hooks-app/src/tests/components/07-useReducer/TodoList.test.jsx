import { shallow } from "enzyme";

import { TodoList } from "../../../components/07-useReducer/TodoList";

import { demoTodos } from "../../fixtures/demoTodos";

describe('TodoList', () => {
  const handleDelete = jest.fn();
  const handleToggle = jest.fn();

  const wrapper = shallow(
    <TodoList
      todos={ demoTodos }
      handleDelete={ handleDelete }
      handleToggle={ handleToggle }
    />
  );

  test('should render', () => {
    expect(wrapper).toMatchSnapshot();
  });

  test('should have 2 <TodoListItem />', () => {
    expect(wrapper.find('TodoListItem').length).toBe(demoTodos.length);
    expect(wrapper.find('TodoListItem').at(0).prop('handleDelete')).toEqual(expect.any(Function));
  });
});
