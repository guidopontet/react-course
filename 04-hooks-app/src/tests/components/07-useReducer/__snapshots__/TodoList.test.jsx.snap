// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`TodoList should render 1`] = `
<ul
  className="list-group list-group-flush"
>
  <TodoListItem
    handleDelete={[MockFunction]}
    handleToggle={[MockFunction]}
    index={0}
    key="1"
    todo={
      Object {
        "done": false,
        "id": 1,
        "text": "Learn React",
      }
    }
  />
  <TodoListItem
    handleDelete={[MockFunction]}
    handleToggle={[MockFunction]}
    index={1}
    key="2"
    todo={
      Object {
        "done": false,
        "id": 2,
        "text": "Learn Redux",
      }
    }
  />
</ul>
`;
